### ISSUE TITLE
Enrich new-session event with the new fields

### ISSUE THREAD

**Issue by zohaib01khan** [hicommonwealth/chain-events#27]
currently, we need additional fields with new-session events for the new validators table record.

- commissionPer
- apr
- uptime
- hasMessage
- eraPoints
- isElected
- toBeElected
- blockCount
- otherTotal
- name

**Comment by drewstone** [comment]
You shouldn't need to get all these values from the `new-session` event. You should be able to get some of these values from other events such as when a new validator record is created.

Also:
- apr is calculated over time, not on `new-session` event.
- era points should be queries and added to the enricher for the `new-session` event
- we do not need `hasMessage` `isElected` `toBeElected` on Validator models. The state should reflect all that info

**Comment by drewstone** [comment]
@jnaviask we should likely add a `utililty.batch` Event to the chain-events repo since I'm not so sure we will catch the internal events of one of one of those.

**Comment by jnaviask** [comment]
Yes -- we should add a `utility.batch` event. What behavior should it have? Perhaps unwrapping and recursively processing all the sub-events?

**Comment by drewstone** [comment]
That makes the most sense to me, we could also use it to check extrinsics inside that event.

**Comment by jnaviask** [comment]
All extrinsics are already run through the chain events system anyway, so that would be part of processing it.

**Comment by zohaib01khan** [comment]
@drewstone  I have updated the details, It is needed for HistoricalValidatorStatistic table when a new validator needs to be added I need that as the Model of the table is https://github.com/hicommonwealth/commonwealth/pull/709/files?file-filters%5B%5D=.ts#diff-36fb28a1746e14fcc324e20adf2750f1

**Comment by drewstone** [comment]
@jnaviask Right but extrinsics inside the utility.batch aren't because they're nested, so we'll have to push those through it again.

**Comment by YameenMalik** [comment]
> You shouldn't need to get all these values from the `new-session` event. You should be able to get some of these values from other events such as when a new validator record is created.

@drewstone Right now we are updating the Validators  table when a new session is triggered and it makes sense as the new-session event is currently enriched us to provide all the Validators and their Exposures. At the moment validator's table is just being updated via new-session event. Is this the expected behavior or something else is expected? There are couple of things which we don't get from new-session event which are mentioned in  validator's table [schema](https://www.notion.so/Move-to-backend-4d995500e32645e1aea7ff8a9ef6a920#6bedb77310694ef5af455d8f42e891e7) which are:
```
controller: AccountId; --string
sessionKeys: AccountId[]; ---string array
```
How shall we go about updating these attributes? Which events/apis can provide us these pieces of information?

Historical Validators Table:
- Apr: Its a function of exposure, reward and commission so I think we need to recompute the APR for every event(reward, slash,  bond/unbond) that updates any of these values.
- era points/ commission: we will enrich new-session to return us these values. Or there any other alternate or better way to get these?
The rest of the values isElected, tobeElected etc... yeah we don't need any enrichment for this.

Thanks.

**Comment by drewstone** [comment]
I would enrich the event to grab the controll and the sessionKeys (controller is most important, sessionKeys less so)

**Comment by drewstone** [comment]
https://github.com/hicommonwealth/substrate/blob/master/frame/staking/src/lib.rs#L981

**Comment by drewstone** [comment]
Getting session keys uses the session module: https://github.com/hicommonwealth/substrate/blob/master/frame/session/src/lib.rs#L415

**Comment by drewstone** [comment]
I have an algorithm in the repo for calculating the APR based on the event data, you should look at that as a reference. I believe it only calculates based on the last most recent 2 events from the DBs, we would want to instead be updating a moving average on these models.

**Comment by drewstone** [comment]
I think enriching on the session event for eraPoints and commission is fine @YameenMalik @HussainMehdi

**Comment by drewstone** [comment]
APR should be re-calculated on `reward` and `slash` events since those are events that update ones stake and exposure. I haven't yet seen event handlers for creating these DB models

**Comment by YameenMalik** [comment]
@drewstone yeah we looked up your APR computation code [here](https://github.com/hicommonwealth/commonwealth/blob/210d71b3daf85c5f7fb17276833bb93fd69dd94c/client/scripts/controllers/chain/substrate/staking.ts#L273) and we intend to use the same function to perform APR computation. You mentioned re-calculating APR on `reward` and `slash` event that I agree upon but since `bond` and `unbond` events effect the exposure of the validator shouldn't we be re-computing APR on these two events as well? 

Event handlers for both `reward` and `slash` are being worked on by @zohaib01khan, the PR is not yet raised as both of these events effect the APR, once that  is done, we will push changes. Event handlers for bond/unbond/new-session all are finalized, zohaib intends to raise a single PR for all the event handlers. Let us know if you intend to see a separate PR for every handler. 

Thanks.
